Implement a cluster-wide max user topic count limit

ballard26 opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR adds a cluster configuration that allows for a maximum number of user created topics to be configured and enforced. It is being enforced in `topics_frontend::do_create_topic` to ensure that the limit is enforced during auto topic creation as well.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes
### Features
* adds the `kafka_topics_max` cluster configuration that can be used to enforce a maximum user topic count cluster-wide.
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

ballard26 committed:
config: add kafka_topics_max cluster config

ballard26 committed:
cluster: enforce topic count limit in topics_frontend

ballard26 committed:
rptest: add ducktape test for topic count limits

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--0439f59a-e3d8-4fbb-91ea-11fc2ceb3088-->
<!--TEST RESULTS-->
<!--72b2fd0a-fb1a-4883-84e9-63e8827109ea-->
<!--TEST RESULTS-->
<!--b9a9c8f0-d206-49ec-80e7-e77e658388e1-->
<!--TEST RESULTS-->
<!--c31a6575-4e39-427f-aa1e-279cdfd46b12-->
<!--TEST RESULTS-->
<!--d9891fbb-2b9a-40ee-9541-249af2217093-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66654">build#66654</a></summary>


| test_class        | test_method                                | test_arguments                                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:------------------|:-------------------------------------------|:--------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| MasterTestSuite   | test_topic_recreation_recovery             |                                                                     | unit        | https://buildkite.com/redpanda/redpanda/builds/66654#0197288e-b21e-4c41-92dd-54411bd3d52e | FAIL          | 0/1      |                                                                                                                                                           |
| DeleteRecordsTest | test_delete_records_concurrent_truncations | {"cloud_storage_enabled": false, "truncate_point": "random_offset"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66654#019728fc-e41d-46dd-979f-33f994ee7a24 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66755">build#66755</a></summary>


| test_class                | test_method                           | test_arguments                                                                                        | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:--------------------------------------|:------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MaintenanceTest           | test_maintenance_sticky               | {"use_rpk": true}                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66755#0197368e-a7d2-4d6d-b823-e165d98b9cb1 | FLAKY         | 20/21    | upstream reliability is '97.36842105263158'. current run reliability is '95.23809523809523'. drift is 2.13033 and the allowed drift is set to 50. The test should PASS |
| PartitionBalancerTest     | test_rack_awareness                   |                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66755#0197368e-a7d5-49a7-8c9a-f4e078274e58 | FLAKY         | 20/21    | upstream reliability is '99.31972789115646'. current run reliability is '95.23809523809523'. drift is 4.08163 and the allowed drift is set to 50. The test should PASS |
| PartitionMoveInterruption | test_cancelling_partition_move_x_core | {"compacted": false, "recovery": "restart_recovery", "replication_factor": 3, "unclean_abort": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66755#019736a1-a0d3-43c0-9d23-e331293f2874 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| DisablingPartitionsTest   | test_disable                          |                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66755#019736a1-a0d2-46cb-b6bb-2b3fa0ebb6d1 | FLAKY         | 16/21    | upstream reliability is '83.7890625'. current run reliability is '76.19047619047619'. drift is 7.59859 and the allowed drift is set to 50. The test should PASS        |
| TopicLimitsTest           | test_limit_auto_creation              |                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66755#0197368e-a7d3-4416-8cf3-1e392d260bf4 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| TopicLimitsTest           | test_limit_auto_creation              |                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66755#019736a1-a0d1-4c26-901d-8f2aa0c5b622 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| TopicLimitsTest           | test_limit_manual_creation            |                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66755#0197368e-a7d4-4456-96d8-d41f303583a2 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| TopicLimitsTest           | test_limit_manual_creation            |                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66755#019736a1-a0d2-46cb-b6bb-2b3fa0ebb6d1 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66928">build#66928</a></summary>


| test_class        | test_method         | test_arguments                                                                                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                               |
|:------------------|:--------------------|:---------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| SaslPlainTest     | test_plain_authn    | {"client_type": 2, "sasl_plain_enabled": true, "scram_type": 1}                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66928#01974119-d8b3-46b4-a245-3412ff5ca532 | FLAKY         | 20/21    | upstream reliability is '99.581589958159'. current run reliability is '95.23809523809523'. drift is 4.34349 and the allowed drift is set to 50. The test should PASS |
| TieredStorageTest | test_tiered_storage | {"cloud_storage_type_and_url_style": [1, "path"], "test_case": {"name": "(TS_Read == True, SpilloverManifestUploaded == True)"}} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66928#01974116-843d-440e-83d1-12facba34be9 | FAIL          | 0/1      |                                                                                                                                                                      |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67067">build#67067</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:-------------------------|:---------------------|:-----------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67067#0197572c-f189-4ec0-8034-264645502241 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67276">build#67276</a></summary>


| test_class                 | test_method                                       | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------|:--------------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionReassignmentsTest | test_add_partitions_with_inprogress_reassignments |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67276#0197659e-0eb8-4020-86f7-032e2c5ba0be | FLAKY         | 17/21    | upstream reliability is '94.01197604790418'. current run reliability is '80.95238095238095'. drift is 13.0596 and the allowed drift is set to 50. The test should PASS |
| RecoveryModeTest           | test_rolling_restart                              |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67276#0197659e-0eba-4bb3-b591-be95b0221188 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<!--/body-->

ballard26 commented:
The unit test failure in the release build don't seem related to this PR.


I'm unable to reproduce the failure running locally using bazel's default scheduling;
```
brandonallard@fedora:~/data/redpanda_3$ bazel test --config=release --runs_per_test=1000 //src/v/kafka/server/tests:topic_recreate_test
INFO: Invocation ID: 39d39c60-771e-48d2-b4b4-37fbd95f4d78
INFO: Analyzed target //src/v/kafka/server/tests:topic_recreate_test (0 packages loaded, 4 targets configured).
INFO: Found 1 test target...
Target //src/v/kafka/server/tests:topic_recreate_test up-to-date:
  bazel-bin/src/v/kafka/server/tests/topic_recreate_test
INFO: Elapsed time: 562.238s, Critical Path: 14.55s
INFO: 1001 processes: 10 action cache hit, 1 internal, 1000 linux-sandbox.
INFO: Build completed successfully, 1001 total actions
//src/v/kafka/server/tests:topic_recreate_test                           PASSED in 3.9s
  Stats over 1000 runs: max = 3.9s, min = 3.2s, avg = 3.4s, dev = 0.1s
```
However, if bazel is forced to schedule more test threads than available cores the issue can be replicated on both this branch and the tip of dev;
```
brandonallard@fedora:~/data/redpanda_3$ 2bazel test --config=release --runs_per_test=1000 //src/v/kafka/server/tests:topic_recreate_test --local_test_jobs=100 --jobs=100
INFO: Invocation ID: 57d3cde4-2be7-4545-b80e-ac732fe473b2
INFO: Analyzed target //src/v/kafka/server/tests:topic_recreate_test (0 packages loaded, 0 targets configured).

FAILED: //src/v/kafka/server/tests:topic_recreate_test (Summary)
      /home/brandonallard/data/.cache/bazel/_bazel_brandonallard/50acd7c106e570e66c6351a7103182f1/execroot/_main/bazel-out/k8-opt/testlogs/src/v/kafka/server/tests/topic_recreate_test/run_39_of_1000/test.log
      /home/brandonallard/data/.cache/bazel/_bazel_brandonallard/50acd7c106e570e66c6351a7103182f1/execroot/_main/bazel-out/k8-opt/testlogs/src/v/kafka/server/tests/topic_recreate_test/run_101_of_1000/test.log
      /home/brandonallard/data/.cache/bazel/_bazel_brandonallard/50acd7c106e570e66c6351a7103182f1/execroot/_main/bazel-out/k8-opt/testlogs/src/v/kafka/server/tests/topic_recreate_test/run_121_of_1000/test.log
      /home/brandonallard/data/.cache/bazel/_bazel_brandonallard/50acd7c106e570e66c6351a7103182f1/execroot/_main/bazel-out/k8-opt/testlogs/src/v/kafka/server/tests/topic_recreate_test/run_216_of_1000/test.log
      /home/brandonallard/data/.cache/bazel/_bazel_brandonallard/50acd7c106e570e66c6351a7103182f1/execroot/_main/bazel-out/k8-opt/testlogs/src/v/kafka/server/tests/topic_recreate_test/run_473_of_1000/test.log
      /home/brandonallard/data/.cache/bazel/_bazel_brandonallard/50acd7c106e570e66c6351a7103182f1/execroot/_main/bazel-out/k8-opt/testlogs/src/v/kafka/server/tests/topic_recreate_test/run_560_of_1000/test.log
      /home/brandonallard/data/.cache/bazel/_bazel_brandonallard/50acd7c106e570e66c6351a7103182f1/execroot/_main/bazel-out/k8-opt/testlogs/src/v/kafka/server/tests/topic_recreate_test/run_705_of_1000/test.log
      /home/brandonallard/data/.cache/bazel/_bazel_brandonallard/50acd7c106e570e66c6351a7103182f1/execroot/_main/bazel-out/k8-opt/testlogs/src/v/kafka/server/tests/topic_recreate_test/run_783_of_1000/test.log
      /home/brandonallard/data/.cache/bazel/_bazel_brandonallard/50acd7c106e570e66c6351a7103182f1/execroot/_main/bazel-out/k8-opt/testlogs/src/v/kafka/server/tests/topic_recreate_test/run_821_of_1000/test.log
      /home/brandonallard/data/.cache/bazel/_bazel_brandonallard/50acd7c106e570e66c6351a7103182f1/execroot/_main/bazel-out/k8-opt/testlogs/src/v/kafka/server/tests/topic_recreate_test/run_917_of_1000/test.log
      /home/brandonallard/data/.cache/bazel/_bazel_brandonallard/50acd7c106e570e66c6351a7103182f1/execroot/_main/bazel-out/k8-opt/testlogs/src/v/kafka/server/tests/topic_recreate_test/run_941_of_1000/test.log
INFO: From Testing //src/v/kafka/server/tests:topic_recreate_test (run 941 of 1000):
stdout (/home/brandonallard/data/.cache/bazel/_bazel_brandonallard/50acd7c106e570e66c6351a7103182f1/execroot/_main/bazel-out/_tmp/actions/stdout-942) 2299297 exceeds maximum size of --experimental_ui_max_stdouterr_bytes=1048576 bytes; skipping
INFO: Found 1 test target...
Target //src/v/kafka/server/tests:topic_recreate_test up-to-date:
  bazel-bin/src/v/kafka/server/tests/topic_recreate_test
INFO: Elapsed time: 67.479s, Critical Path: 9.97s
INFO: 1001 processes: 3 action cache hit, 1 internal, 1000 linux-sandbox.
INFO: Build completed, 1 test FAILED, 1001 total actions
//src/v/kafka/server/tests:topic_recreate_test                           FAILED in 11 out of 1000 in 9.7s
  Stats over 1000 runs: max = 9.7s, min = 4.1s, avg = 6.5s, dev = 0.8s
```


The actual failure is a timeout;
```
unknown location(0): fatal error: in "test_topic_recreation_recovery": seastar::timed_out_error: timedout
```
from
```
wait_until_topic_status(tp: test_tp, ec: kafka::error_code::none).get();
```
at `src/v/kafka/server/tests/topic_recreate_test.cc:198`. Increasing the timeout from 3s to 5+ minutes doesn't allow the test to pass so there is some bug here. Looking at the test reilability on `dev` over the last 28 days(https://buildkite.com/organizations/redpanda/analytics/suites/redpanda-unit/tests/595ea415-947a-899d-9f63-8d4da4b98215?period=28days&tags=scm.branch%3Adev) we can see that it's periodically failed in the same way as well.

StephanDollberg commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -841,6 +843,18 @@ ss::future<topic_result> topics_frontend::do_create_topic(
+    auto max_user_topics_opt = _max_user_topics();
+    if (
+      max_user_topics_opt.has_value()
+      && _topics.local().all_topics_count() > *max_user_topics_opt
```
So I guess there is no way to easily get the user topic count?

ballard26 replied:
I could do something like `std::min(_topics.local().all_topics_count() - 5, 0)` to account for ~5 non-user topics RP tends to create. Or for an exact solution I can get a list of the topic names, filter out non-user topics, then count that. However, since the non-user topic count will always be significantly smaller than the user topic count I figured it wasn't worth the added complexity.

StephanDollberg replied:
I guess we should do something at least. As otherwise it could get odd.

ballard26 replied:
Fair enough, added a simple approximation in.

michael-redpanda commented on src/v/config/configuration.cc:
```diff
@@ -1714,6 +1714,13 @@ configuration::configuration()
+      "kafka_topics_max",
+      "Maximum number of Kafka topics per cluster. If `null`, then no limit is "
+      "enforced.",
+      {.needs_restart = needs_restart::no, .visibility = visibility::user},
```
question: Should this be user visible or tunable?  What's the use case for self-hosted customers?

ballard26 replied:
Topics add 60+ metric series per-topic and are not aggregated in most cases. The use of this config for cloudv2 and self-hosted is mainly a safety measure to avoid more metric series than a given Prometheus instance can handle.

ballard26 replied:
As for the visibility, I have no strong opinion about it being user or tunable in this case.

michael-redpanda reviewed (commented):
lgtm, just wondering what's driving the need for this vs our partition limits?

ballard26 committed:
model: define an array of non-user topics in the kafka namespace

Note that the switch to `std::find` is both more succinct and efficient;
test                       iterations      median         mad      min         max      allocs       tasks        inst      cycles
sb.is_user_topic                36621    26.595us     9.306ns 26.578us    26.605us       0.000       0.000    560360.1    140431.5
sb.is_user_topic_std_find       46706    20.759us     6.619ns 20.753us    20.774us       0.000       0.000    280360.0    110420.8

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/66755-->
<!--9fdee7e4-3f84-49e7-937a-486c02a174e4-->
<!--713ed044-2627-44eb-ad62-6f3060716ecc-->
<!--a5f11461-4c8b-46a2-aa41-e9530e72e411-->
<!--9b8aac30-0148-4a16-b397-4924f9c6ba7a-->
<!--title-->
## Retry command for [Build#66755](https://buildkite.com/redpanda/redpanda/builds/66755)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/topic_limits_test.py::TopicLimitsTest.test_limit_auto_creation
tests/rptest/tests/topic_limits_test.py::TopicLimitsTest.test_limit_manual_creation
```
<!--/body-->

dotnwat commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -841,6 +843,26 @@ ss::future<topic_result> topics_frontend::do_create_topic(
+        if (
+          user_topic_count > *max_user_topics_opt
+          && model::is_user_topic(tp_ns)) {
+            co_return make_error_result(
+              assignable_config.cfg.tp_ns,
+              make_error_code(errc::topic_operation_error));
```
since the returned error is very generic, should we log a message about why creation failed? also, `kafka::protocol::creatable_topic_response` looks like it contains a free-form string `error_message` that could provide more context to the client?

ballard26 replied:
Good idea, will add a message.

ballard26 commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -841,6 +843,27 @@ ss::future<topic_result> topics_frontend::do_create_topic(
+                                - model::non_user_topics.size();
+        if (
+          user_topic_count > *max_user_topics_opt
+          && model::is_user_topic(tp_ns)) {
```
I mainly left it as the final check as its the most expensive. It may hurt readability a bit, however, I wouldn't expect this limit to be hit in most cases and therefore wanted its addition to have the least overhead possible.

ballard26 commented on tests/rptest/tests/topic_limits_test.py:
```diff
@@ -0,0 +1,88 @@
+        super(TopicLimitsTest, self).__init__(
+            test_context=test_context,
+            num_brokers=3,
+            extra_rp_conf={'auto_create_topics_enabled': False})
```
Mainly to be explicit about the test changing it and what it's initial value is. Will remove if you feel that it isn't needed though.

ballard26 commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -841,6 +843,27 @@ ss::future<topic_result> topics_frontend::do_create_topic(
     if (!units) {
         co_return make_error_result(assignable_config.cfg.tp_ns, units.error());
     }
+
```
Good point, will move this check towards the beginning of the function.

ballard26 commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -841,6 +843,27 @@ ss::future<topic_result> topics_frontend::do_create_topic(
+
+    // Enforce cluster-wide topic count limit only on user created topics.
+    auto max_user_topics_opt = _max_user_topics();
+    if (max_user_topics_opt.has_value()) {
```
Nice, forgot about this. Will switch over.

ballard26 commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -841,6 +843,27 @@ ss::future<topic_result> topics_frontend::do_create_topic(
+                                  model::non_user_topics.size())
+                                - model::non_user_topics.size();
+        if (
+          user_topic_count > *max_user_topics_opt
```
Right you are, switching to `>=`

ballard26 commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -841,6 +843,27 @@ ss::future<topic_result> topics_frontend::do_create_topic(
+        if (
+          user_topic_count > *max_user_topics_opt
+          && model::is_user_topic(tp_ns)) {
+            co_return make_error_result(
```
Good point, adding a log message.

ballard26 commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -841,6 +843,27 @@ ss::future<topic_result> topics_frontend::do_create_topic(
+          && model::is_user_topic(tp_ns)) {
+            co_return make_error_result(
+              assignable_config.cfg.tp_ns,
+              make_error_code(errc::topic_operation_error),
```
No it won't be the exact same error code. Mainly since for partitions the Kafka API has specific error codes for why partition allocation can fail, but nothing for general topic creation. I.e:
``` 
topic_invalid_partitions_core_limit,
topic_invalid_partitions_memory_limit,
topic_invalid_partitions_fd_limit,
topic_invalid_partitions_decreased
 ```
I did provide a custom error message to explain what exactly happened to the client. But the error code itself is rather general and I couldn't find a more specific one that made sense out of what is available. There may be some argument to actually use one of the partition error messages as it conveys to clients that the cluster is resource constrained, but I'm unsure about it.

ballard26 commented on src/v/config/configuration.cc:
```diff
@@ -1714,6 +1714,13 @@ configuration::configuration()
+  , kafka_topics_max(
+      *this,
+      "kafka_topics_max",
+      "Maximum number of Kafka topics per cluster. If `null`, then no limit is "
```
Switching to your suggestion

ballard26 commented on src/v/model/namespace.h:
```diff
@@ -84,12 +84,19 @@ inline const model::topic datalake_coordinator_topic("datalake_coordinator");
-           && tp_ns.tp != schema_registry_internal_tp.topic
-           && tp_ns.tp != kafka_audit_logging_topic
-           && tp_ns.tp != transform_log_internal_topic;
+           && std::find(
```
Heh, yeah, will need to stop reaching for `std::find`. Switching to `std::ranges::contains`

micheleRP commented on src/v/config/configuration.cc:
```diff
@@ -1714,6 +1714,13 @@ configuration::configuration()
+  , kafka_topics_max(
+      *this,
+      "kafka_topics_max",
+      "Maximum number of Kafka user topics which can be created. If `null`, "
```
```suggestion
      "Maximum number of Kafka user topics that can be created. If `null`, "
```

ballard26 replied:
Made your requested change

ballard26 commented on tests/rptest/tests/topic_limits_test.py:
```diff
@@ -0,0 +1,88 @@
+        assert self._within(failed_attempts, topic_limit, 0.25)
+
+        topic_count = len(self.client().describe_topics())
+        assert self._within(topic_count, topic_limit, 0.25)
```
Added the additional test you recommended.
